SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp

```diff
@@ -0,0 +1,1833 @@
+		ApplyComponentUpdate(ComponentId, Update, *TargetObject, *Channel, /* bIsHandover */ false);
+	}
+	else if (Category == SCHEMA_Handover)
+	{
+		// SCOPE_CYCLE_COUNTER(STAT_ReceiverApplyHandover);
+		if (!NetDriver->IsServer())
+		{
+			UE_LOG(LogActorSystem, Verbose, TEXT("Entity: %d Component: %d - Skipping Handover component because we're a client."),
+				   EntityId, ComponentId);
+			return;
+		}
+
+		ApplyComponentUpdate(ComponentId, Update, *TargetObject, *Channel, /* bIsHandover */ true);
+	}
+	else
+	{
+		UE_LOG(LogActorSystem, Verbose,
+			   TEXT("Entity: %d Component: %d - Skipping because it's an empty component update from an RPC component. (most likely as a "
+					"result of gaining authority)"),
+			   EntityId, ComponentId);
+	}
+}
+
+void ActorSystem::ComponentRemoved(const Worker_EntityId EntityId, const Worker_ComponentId ComponentId) const
+{
+	// Early out if this isn't a generated component.
+	if (SemanticActorComponents.Contains(ComponentId))
+	{
+		return;
+	}
+
+	if (AActor* Actor = Cast<AActor>(NetDriver->PackageMap->GetObjectFromEntityId(EntityId).Get()))
+	{
+		FUnrealObjectRef ObjectRef(EntityId, ComponentId);
+		if (ComponentId == SpatialConstants::DORMANT_COMPONENT_ID)
+		{
+			GetOrRecreateChannelForDormantActor(Actor, EntityId);
+		}
+		else if (UObject* Object = NetDriver->PackageMap->GetObjectFromUnrealObjectRef(ObjectRef).Get())
+		{
+			if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(EntityId))
+			{
+				TWeakObjectPtr<UObject> WeakPtr(Object);
+				Channel->OnSubobjectDeleted(ObjectRef, Object, WeakPtr);
+
+				Actor->OnSubobjectDestroyFromReplication(Object);
+
+				Object->PreDestroyFromReplication();
+				Object->MarkPendingKill();
+
+				NetDriver->PackageMap->RemoveSubobject(FUnrealObjectRef(EntityId, ComponentId));
+			}
+		}
+	}
+}
+
+void ActorSystem::EntityAdded(const Worker_EntityId EntityId)
+{
+	ReceiveActor(EntityId);
+	for (const auto& AuthoritativeComponentSet : SubView->GetView()[EntityId].Authority)
+	{
+		AuthorityGained(EntityId, AuthoritativeComponentSet);
+	}
+}
+
+void ActorSystem::EntityRemoved(const Worker_EntityId EntityId)
+{
+	// SCOPE_CYCLE_COUNTER(STAT_ReceiverRemoveEntity);
+
+	RemoveActor(EntityId);
+
+	// Stop tracking if the entity was deleted as a result of deleting the actor during creation.
+	// This assumes that authority will be gained before interest is gained and lost.
+	const int32 RetiredActorIndex = EntitiesToRetireOnAuthorityGain.IndexOfByPredicate([EntityId](const DeferredRetire& Retire) {
+		return EntityId == Retire.EntityId;
+	});
+	if (RetiredActorIndex != INDEX_NONE)
+	{
+		EntitiesToRetireOnAuthorityGain.RemoveAtSwap(RetiredActorIndex);
+	}
+
+	if (NetDriver->IsServer())
+	{
+		// Check to see if we are removing a system entity for a worker connection. If so clean up the ClientConnection to delete any and
+		// all actors for this connection's controller.
+		if (FString* WorkerName = WorkerConnectionEntities.Find(EntityId))
```

Comment:
Should be moved somewhere else, this would break Client disconnection detection otherwise.

Reply:
yeah, it should be. A new subsystem for system entities is a good idea, and we can probably use the system component as a tag. Will add this as a follow up ticket